File: internal/dao/jinzhu/dbr/course.go
// Copyright 2022 ROC. All rights reserved.
// Use of this source code is governed by a MIT style
// license that can be found in the LICENSE file.

package dbr

import (
	"time"

	"gorm.io/gorm"
)

// CourseGroup 课程分组(管理员动态维护)
type CourseGroup struct {
	*Model
	Name string `json:"name"`
	Sort int    `json:"sort"`
}

// CourseGroupFormated 课程分组(含课程数)
type CourseGroupFormated struct {
	ID          int64  `json:"id"`
	Name        string `json:"name"`
	Sort        int    `json:"sort"`
	CourseCount int64  `json:"course_count"`
}

func (g *CourseGroup) Format(courseCount int64) *CourseGroupFormated {
	if g.Model == nil {
		return &CourseGroupFormated{}
	}
	return &CourseGroupFormated{
		ID:          g.Model.ID,
		Name:        g.Name,
		Sort:        g.Sort,
		CourseCount: courseCount,
	}
}

func (g *CourseGroup) Create(db *gorm.DB) (*CourseGroup, error) {
	err := db.Create(&g).Error
	return g, err
}

func (g *CourseGroup) Get(db *gorm.DB) (*CourseGroup, error) {
	var group CourseGroup
	if g.Model == nil || g.ID <= 0 {
		return nil, gorm.ErrRecordNotFound
	}
	err := db.Where("id = ? AND is_del = ?", g.ID, 0).First(&group).Error
	if err != nil {
		return &group, err
	}
	return &group, nil
}

func (g *CourseGroup) List(db *gorm.DB) ([]*CourseGroup, error) {
	var groups []*CourseGroup
	err := db.Where("is_del = ?", 0).Order("sort, id").Find(&groups).Error
	return groups, err
}

func (g *CourseGroup) Update(db *gorm.DB) error {
	return db.Model(g).Where("id = ?", g.Model.ID).Updates(map[string]any{
		"name": g.Name,
		"sort": g.Sort,
	}).Error
}

// Delete 分组硬删除(调用方需保证分组下无课程)
func (g *CourseGroup) Delete(db *gorm.DB) error {
	return db.Unscoped().Where("id = ?", g.Model.ID).Delete(&CourseGroup{}).Error
}

// Course 课程: 无状态流转, 创建即上线; 删除为硬删除
type Course struct {
	*Model
	GroupID      int64  `json:"group_id"`
	TeacherID    int64  `json:"teacher_id"`
	Title        string `json:"title"`
	Intro        string `json:"intro"`
	VideoURL     string `json:"video_url"`
	Cover        string `json:"cover"`
	PlayCount    int64  `json:"play_count"`
	CommentCount int64  `json:"comment_count"`
}

// CourseFormated 课程输出(含分组名与老师信息)
type CourseFormated struct {
	ID           int64         `json:"id"`
	GroupID      int64         `json:"group_id"`
	GroupName    string        `json:"group_name"`
	TeacherID    int64         `json:"teacher_id"`
	Teacher      *UserFormated `json:"teacher"`
	Title        string        `json:"title"`
	Intro        string        `json:"intro"`
	Cover        string        `json:"cover"`
	PlayCount    int64         `json:"play_count"`
	CommentCount int64         `json:"comment_count"`
	CreatedOn    int64         `json:"created_on"`
}

func (c *Course) Format() *CourseFormated {
	if c.Model == nil {
		return &CourseFormated{}
	}
	return &CourseFormated{
		ID:           c.Model.ID,
		GroupID:      c.GroupID,
		TeacherID:    c.TeacherID,
		Teacher:      &UserFormated{},
		Title:        c.Title,
		Intro:        c.Intro,
		Cover:        c.Cover,
		PlayCount:    c.PlayCount,
		CommentCount: c.CommentCount,
		CreatedOn:    c.CreatedOn,
	}
}

func (c *Course) Create(db *gorm.DB) (*Course, error) {
	err := db.Create(&c).Error
	return c, err
}

func (c *Course) Get(db *gorm.DB) (*Course, error) {
	var course Course
	if c.Model == nil || c.ID <= 0 {
		return nil, gorm.ErrRecordNotFound
	}
	err := db.Where("id = ? AND is_del = ?", c.ID, 0).First(&course).Error
	if err != nil {
		return &course, err
	}
	return &course, nil
}

func (c *Course) Update(db *gorm.DB) error {
	return db.Model(c).Where("id = ?", c.Model.ID).Updates(map[string]any{
		"group_id":   c.GroupID,
		"teacher_id": c.TeacherID,
		"title":      c.Title,
		"intro":      c.Intro,
		"video_url":  c.VideoURL,
		"cover":      c.Cover,
	}).Error
}

// DeleteUnscoped 课程硬删除(评论等关联数据由调用方在同一事务内清理)
func (c *Course) DeleteUnscoped(db *gorm.DB) error {
	return db.Unscoped().Where("id = ?", c.Model.ID).Delete(&Course{}).Error
}

// CourseComment 课程评论: 审核口径与帖子评论一致
type CourseComment struct {
	*Model
	CourseID    int64      `json:"course_id"`
	UserID      int64      `json:"user_id"`
	IP          string     `json:"ip"`
	IPLoc       string     `json:"ip_loc"`
	ReplyCount  int32      `json:"reply_count"`
	AuditStatus PostAuditT `json:"audit_status"`
}

type CourseCommentFormated struct {
	ID          int64                         `json:"id"`
	CourseID    int64                         `json:"course_id"`
	UserID      int64                         `json:"user_id"`
	User        *UserFormated                 `json:"user"`
	Contents    []*CourseCommentContent       `json:"contents"`
	Replies     []*CourseCommentReplyFormated `json:"replies"`
	IPLoc       string                        `json:"ip_loc"`
	ReplyCount  int32                         `json:"reply_count"`
	AuditStatus PostAuditT                    `json:"audit_status"`
	CreatedOn   int64                         `json:"created_on"`
	ModifiedOn  int64                         `json:"modified_on"`
}

func (c *CourseComment) Format() *CourseCommentFormated {
	if c.Model == nil {
		return &CourseCommentFormated{}
	}
	return &CourseCommentFormated{
		ID:          c.Model.ID,
		CourseID:    c.CourseID,
		UserID:      c.UserID,
		User:        &UserFormated{},
		Contents:    []*CourseCommentContent{},
		Replies:     []*CourseCommentReplyFormated{},
		IPLoc:       c.IPLoc,
		ReplyCount:  c.ReplyCount,
		AuditStatus: c.AuditStatus,
		CreatedOn:   c.CreatedOn,
		ModifiedOn:  c.ModifiedOn,
	}
}

func (c *CourseComment) Create(db *gorm.DB) (*CourseComment, error) {
	err := db.Create(&c).Error
	return c, err
}

func (c *CourseComment) Get(db *gorm.DB) (*CourseComment, error) {
	var comment CourseComment
	if c.Model == nil || c.ID <= 0 {
		return nil, gorm.ErrRecordNotFound
	}
	err := db.Where("id = ? AND is_del = ?", c.ID, 0).First(&comment).Error
	if err != nil {
		return &comment, err
	}
	return &comment, nil
}

// Delete 用户删除评论(软删, 与帖子评论一致)
func (c *CourseComment) Delete(db *gorm.DB) error {
	return db.Model(c).Where("id = ?", c.Model.ID).Updates(map[string]any{
		"deleted_on": time.Now().Unix(),
		"is_del":     1,
	}).Error
}

// CourseCommentContent 课程评论内容(分块)
type CourseCommentContent struct {
	*Model
	CommentID int64        `json:"comment_id"`
	UserID    int64        `json:"user_id"`
	Content   string       `json:"content"`
	Type      PostContentT `json:"type"`
	Sort      int64        `json:"sort"`
}

func (c *CourseCommentContent) Create(db *gorm.DB) (*CourseCommentContent, error) {
	err := db.Create(&c).Error
	return c, err
}

// CourseCommentReply 课程评论回复(楼中楼)
type CourseCommentReply struct {
	*Model
	CommentID   int64      `json:"comment_id"`
	UserID      int64      `json:"user_id"`
	AtUserID    int64      `json:"at_user_id"`
	Content     string     `json:"content"`
	IP          string     `json:"ip"`
	IPLoc       string     `json:"ip_loc"`
	AuditStatus PostAuditT `json:"audit_status"`
}

type CourseCommentReplyFormated struct {
	ID          int64         `json:"id"`
	CommentID   int64         `json:"comment_id"`
	UserID      int64         `json:"user_id"`
	User        *UserFormated `json:"user"`
	AtUserID    int64         `json:"at_user_id"`
	AtUser      *UserFormated `json:"at_user"`
	Content     string        `json:"content"`
	IPLoc       string        `json:"ip_loc"`
	AuditStatus PostAuditT    `json:"audit_status"`
	CreatedOn   int64         `json:"created_on"`
	ModifiedOn  int64         `json:"modified_on"`
}

func (c *CourseCommentReply) Format() *CourseCommentReplyFormated {
	if c.Model == nil {
		return &CourseCommentReplyFormated{}
	}
	return &CourseCommentReplyFormated{
		ID:          c.ID,
		CommentID:   c.CommentID,
		UserID:      c.UserID,
		User:        &UserFormated{},
		AtUserID:    c.AtUserID,
		AtUser:      &UserFormated{},
		Content:     c.Content,
		IPLoc:       c.IPLoc,
		AuditStatus: c.AuditStatus,
		CreatedOn:   c.CreatedOn,
		ModifiedOn:  c.ModifiedOn,
	}
}

func (c *CourseCommentReply) Create(db *gorm.DB) (*CourseCommentReply, error) {
	err := db.Create(&c).Error
	return c, err
}

func (c *CourseCommentReply) Get(db *gorm.DB) (*CourseCommentReply, error) {
	var reply CourseCommentReply
	if c.Model == nil || c.ID <= 0 {
		return nil, gorm.ErrRecordNotFound
	}
	err := db.Where("id = ? AND is_del = ?", c.ID, 0).First(&reply).Error
	if err != nil {
		return &reply, err
	}
	return &reply, nil
}

// Delete 用户删除回复(软删, 与帖子回复一致)
func (c *CourseCommentReply) Delete(db *gorm.DB) error {
	return db.Model(c).Where("id = ? AND is_del = ?", c.Model.ID, 0).Updates(map[string]any{
		"deleted_on": time.Now().Unix(),
		"is_del":     1,
	}).Error
}

File: web/src/api/course.ts
/**
 * 课程模块 API(手写封装, 与 @/api/post 同一风格; DELETE 动词 Proxy 客户端不支持故全部走这里)
 */

import { request } from '@/utils/request';

export interface CourseUserBrief {
  id: number;
  username: string;
  nickname: string;
  avatar: string;
}

export interface CourseGroup {
  id: number;
  name: string;
  sort: number;
  course_count: number;
}

export interface CourseItem {
  id: number;
  group_id: number;
  group_name: string;
  teacher_id: number;
  teacher: CourseUserBrief;
  title: string;
  intro: string;
  cover: string;
  play_count: number;
  comment_count: number;
  created_on: number;
}

export interface CourseCommentContent {
  id: number;
  comment_id: number;
  user_id: number;
  content: string;
  type: number;
  sort: number;
}

export interface CourseReply {
  id: number;
  comment_id: number;
  user_id: number;
  user: CourseUserBrief;
  at_user_id: number;
  at_user: CourseUserBrief;
  content: string;
  ip_loc: string;
  audit_status: number;
  created_on: number;
  modified_on: number;
}

export interface CourseComment {
  id: number;
  course_id: number;
  user_id: number;
  user: CourseUserBrief;
  contents: CourseCommentContent[];
  replies: CourseReply[];
  ip_loc: string;
  reply_count: number;
  audit_status: number;
  created_on: number;
  modified_on: number;
}

export interface PageResp<T> {
  list: T[];
  pager: {
    page: number;
    page_size: number;
    total_rows: number;
  };
}

export interface UploadCredential {
  mode: 'direct' | 'proxy';
  host?: string;
  access_key_id?: string;
  policy?: string;
  signature?: string;
  key?: string;
  expire?: number;
}

/** 课程分组列表(含各组课程数) */
export const getCourseGroups = (): Promise<{ groups: CourseGroup[] }> => {
  return request({ method: 'get', url: '/v1/course/groups' });
};

/** 课程列表: group_id过滤分组, keyword搜索标题/简介 */
export const getCourseList = (params: {
  group_id?: number;
  keyword?: string;
  page: number;
  page_size: number;
}): Promise<PageResp<CourseItem>> => {
  return request({ method: 'get', url: '/v1/course/list', params });
};

/** 课程详情 */
export const getCourse = (params: {
  id: number;
}): Promise<{ course: CourseItem }> => {
  return request({ method: 'get', url: '/v1/course', params });
};

/** 课程评论列表 */
export const getCourseComments = (params: {
  id: number;
  page: number;
  page_size: number;
}): Promise<PageResp<CourseComment>> => {
  return request({ method: 'get', url: '/v1/course/comments', params });
};

/** 课程视频签名播放地址 */
export const getCourseVideo = (params: {
  id: number;
}): Promise<{ signed_url: string }> => {
  return request({ method: 'get', url: '/v1/course/video', params });
};

/** 播放计数(每次播放+1) */
export const playCourse = (data: {
  id: number;
}): Promise<{ play_count: number }> => {
  return request({ method: 'post', url: '/v1/course/play', data });
};

/** 发布课程评论(无角色用户进入审核) */
export const createCourseComment = (data: {
  course_id: number;
  contents: { content: string; type: number; sort: number }[];
  users: string[];
}): Promise<CourseComment> => {
  return request({ method: 'post', url: '/v1/course/comment', data });
};

/** 删除课程评论(本人或管理员) */
export const deleteCourseComment = (data: { id: number }): Promise<unknown> => {
  return request({ method: 'delete', url: '/v1/course/comment', data });
};

/** 回复课程评论 */
export const createCourseCommentReply = (data: {
  comment_id: number;
  at_user_id: number;
  content: string;
}): Promise<CourseReply> => {
  return request({ method: 'post', url: '/v1/course/comment/reply', data });
};

/** 删除课程回复(本人或管理员) */
export const deleteCourseCommentReply = (data: {
  id: number;
}): Promise<unknown> => {
  return request({ method: 'delete', url: '/v1/course/comment/reply', data });
};

// ===== 管理接口(管理员/运维) =====

export const createCourseGroup = (data: {
  name: string;
  sort: number;
}): Promise<CourseGroup> => {
  return request({ method: 'post', url: '/v1/admin/course/group', data });
};

export const updateCourseGroup = (data: {
  id: number;
  name: string;
  sort: number;
}): Promise<unknown> => {
  return request({ method: 'post', url: '/v1/admin/course/group/update', data });
};

export const deleteCourseGroup = (data: { id: number }): Promise<unknown> => {
  return request({ method: 'post', url: '/v1/admin/course/group/delete', data });
};

export const createCourse = (data: {
  group_id: number;
  teacher_id: number;
  title: string;
  intro: string;
  video: string;
  cover: string;
}): Promise<CourseItem> => {
  return request({ method: 'post', url: '/v1/admin/course', data });
};

export const updateCourse = (data: {
  id: number;
  group_id: number;
  teacher_id: number;
  title: string;
  intro: string;
  video?: string;
  cover?: string;
}): Promise<unknown> => {
  return request({ method: 'post', url: '/v1/admin/course/update', data });
};

export const deleteCourse = (data: { id: number }): Promise<unknown> => {
  return request({ method: 'post', url: '/v1/admin/course/delete', data });
};

/** 获取视频上传凭证(AliOSS返回直传policy, 其他返回proxy) */
export const getCourseUploadCredential = (params: {
  ext: string;
}): Promise<UploadCredential> => {
  return request({
    method: 'get',
    url: '/v1/admin/course/upload-credential',
    params,
  });
};
